gtk: support NixOS standalone

danth opened the pull request:
This branch (work in progress) adds support for GTK at the NixOS level.

Currently the theme works on non-`libadwaita` apps, and on `libadwaita` apps when `$GTK_THEME` is set. I've added a patch to allow `libadwaita` to load the theme by default, but I've not had time to test this yet, since it causes most apps on my system to be built from source. The patch is necessary since `$GTK_THEME` is meant for debugging purposes, so it takes priority over Home Manager and any imperative settings.

The reason why `libadwaita` does not load themes by default is because they create additional work for upstream maintainers who never intended for their app to be themed. As a general note, we might want to add a warning in our README about this, asking users to disable Stylix before sending any bug reports upstream.

Before this is merged, it seems like a good idea to split `libadwaita` into a separate target, so that users can disable the overlay if none of their apps need it.

Later, I believe the Home Manager module could be refactored to use the same theme package I've created here.

---

<!--
Unless otherwise specified, the following checkboxes are not mandatory, but
drastically accelerate the reviewing and merging process of this PR.
-->
- [x] <!-- MANDATORY --> I certify that I have the right to submit this contribution under the [MIT license](https://github.com/nix-community/stylix/blob/master/LICENSE)
- [x] Commit messages adhere to [Stylix commit conventions](https://nix-community.github.io/stylix/commit_convention.html)
- [ ] Theming changes adhere to the [Stylix style guide](https://nix-community.github.io/stylix/styling.html)
- [ ] Changes have been [tested locally](https://nix-community.github.io/stylix/modules.html#development-setup)
- [ ] Changes have been [tested in testbeds](https://nix-community.github.io/stylix/testbeds.html)
- [x] Each commit in this PR is suitable for backport to the current stable branch

danth committed:
gtk: support NixOS standalone

This adds support for GTK apps at the system level, in addition to the
existing Home Manager support.

This currently only works for non-libadwaita apps because libadwaita
overrides the theme name [1] for compatibility reasons.

[1]:
https://gitlab.gnome.org/GNOME/libadwaita/-/blob/43192bc3e2870a142e57f971b00b60d69abc9492/src/adw-style-manager.c#L492-495

danth committed:
gtk: patch libadwaita to allow custom theme loading

This adds support for libadwaita apps at the system level, in addition
to the existing Home Manager support.

0xda157 commented on modules/gtk/overlay.nix:
```diff
@@ -0,0 +1,17 @@
+      (
+        config.stylix.enable
+        && config.stylix.targets ? gtk
+        && config.stylix.targets.gtk.enable
```
```suggestion
        && config.stylix.targets.gtk.enable or false
```

0xda157 commented on modules/gtk/nixos.nix:
```diff
@@ -1,7 +1,119 @@
+          extension = ".css";
+        };
+
+        theme =
```
this seem unnecessarily to have in a `let in` block, why not just inline it?

danth replied:
I'm planning to move it to a separate file so that we can reuse it for Home Manager.

> Later, I believe the Home Manager module could be refactored to use the same theme package I've created here.

danth commented on modules/gtk/overlay.nix:
Yes, that makes sense. I opened https://github.com/nix-community/stylix/discussions/2112 with some more ideas.

Although note that this patch is independent of the theme used, so it should be available from the `nix-community` binary cache, at least for packages that are used in one of our testbeds.

danth committed:
libadwaita: split into a separate module

This makes it possible to disable the libadwaita patch without disabling
GTK as a whole.

danth committed:
libadwaita: add testbed

danth committed:
gtk: move the theme package to a separate file

This is in preparation for unifying the implementation of our NixOS and
Home Manager modules.

trueNAHO commented on modules/gtk/meta.nix:
```diff
@@ -1,6 +1,28 @@
+    loading custom themes.
+
+    > [!WARNING]
+    > Theming is [explicitly unsupported](https://stopthemingmy.app/) by many
```
Nitpick to align with commit bba3152bd28b ("treewide: standardize URL format by removing trailing slashes (#1566)"):


```suggestion
    > Theming is [explicitly unsupported](https://stopthemingmy.app) by many
```

trueNAHO commented on modules/libadwaita/meta.nix:
```diff
@@ -0,0 +1,21 @@
+    compiled from source.
+
+    > [!WARNING]
+    > Theming is [explicitly unsupported](https://stopthemingmy.app/) by many
```
Nitpick to align with commit bba3152bd28b ("treewide: standardize URL format by removing trailing slashes (#1566)"):


```suggestion
    > Theming is [explicitly unsupported](https://stopthemingmy.app) by many
```

trueNAHO reviewed (changes requested):
This PR looks very nice :)

> I've not tested Home Manager standalone as of yet.

Considering the `or false` fallback in

https://github.com/nix-community/stylix/blob/bb641234fbc3f73a0a1092157377597cec9b8a64/modules/libadwaita/overlay.nix#L7

I believe none of the added code should be triggered on standalone Home Manager. Supporting standalone Home Manager could be done in a follow-up PR:

> Updating the Home Manager module to use the new theme package can be done in a separate PR

On a side note, unsure which of "NixOS standalone" and "standalone NixOS" is more "correct", as they imply different hierarchies.

danth commented on modules/gtk/theme.nix:
```diff
@@ -0,0 +1,85 @@
+  runCommandLocal,
+}:
+
+runCommandLocal "stylix-gtk-theme"
```
Yes, I will implement something along these lines so that the INI files are generated from attribute sets.

danth commented on modules/libadwaita/meta.nix:
```diff
@@ -0,0 +1,21 @@
+    the standard Adwaita theme.
+
+    This module patches the source code to permit loading other themes. It's
+    disabled by default because this change causes many applications to be
```
Conceptually I agree with this. When this module's Home Manager counterpart is ported over to the new theme, it would also avoid disabling a feature which was previously enabled by default.

However from a user experience standpoint, compiling a huge chunk of the system from scratch is not ideal, especially when it would be required for someone just quickly trying Stylix for the first time. Also, NixOS + Home Manager seems to be the most popular combination and with the current state of the module, libadwaita is themed in that case even with this disabled.

Overall I'm not really sure which is the best option, but would like to avoid blocking this PR on #2112 if possible.

danth replied:
I was actually not aware of `stylix.overlays.enable` either. Knowing that, it makes sense for this target to follow the standard behaviour.

If heavy rebuilds become an issue, we can change the default for `stylix.overlays.enable` instead, to ensure the decision is applied consistently across all relevant modules.

One thing I'd note is that it's presumably possible to set `stylix.targets.libadwaita.enable = true` and `stylix.overlays.enable = false`, in which case the target would appear to be enabled but have no visible effect. Should we print a warning in that case?

trueNAHO commented on modules/libadwaita/nixos.nix:
```diff
@@ -0,0 +1,3 @@
+{ mkTarget, ... }:
+
+mkTarget { autoEnable = false; }
```
The `mkTarget` module itself does not need to be specifically disabled because the underlying overlay is already guarded by `stylix.overlays.enable`:

```suggestion
mkTarget { }
```

Not respecting `stylix.overlays.enable` is confusing and inconsistent.

danth committed:
gtk: replace Bash script with `writeText` and `linkFarm`

danth commented:
Build failures are:

- MPV assertion which I believe is unrelated to this PR
- For some reason Chromium is being compiled from source, which I did not expect
